File: examples/java/VectorStoreExample.java
package examples;

import com.mongodb.client.MongoDatabase;
import dev.langchain4j.data.document.Metadata;
import dev.langchain4j.data.embedding.Embedding;
import dev.langchain4j.data.segment.TextSegment;
import dev.langchain4j.store.embedding.EmbeddingMatch;
import dev.langchain4j.store.embedding.EmbeddingSearchRequest;
import dev.langchain4j.store.embedding.filter.comparison.IsEqualTo;
import io.documentdb.langchain4j.DocumentDBEmbeddingStore;
import io.documentdb.langchain4j.DocumentDBEmbeddingStoreConfig;
import io.documentdb.langchain4j.DocumentDBSimilarity;
import io.documentdb.langchain4j.DocumentDBVectorIndexKind;
import java.util.List;
import java.util.Map;

public final class VectorStoreExample {
    private VectorStoreExample() {}

    public static List<EmbeddingMatch<TextSegment>> run(
            MongoDatabase database, float[] documentVector, float[] queryVector) {
        var store = new DocumentDBEmbeddingStore(
                DocumentDBEmbeddingStoreConfig.builder()
                        .database(database)
                        .collectionName("documents")
                        .dimensions(documentVector.length)
                        .indexName("embedding_vector")
                        .indexKind(DocumentDBVectorIndexKind.DISKANN)
                        .similarity(DocumentDBSimilarity.COS)
                        .diskAnnMaxDegree(32)
                        .diskAnnBuildCandidates(50)
                        .diskAnnSearchCandidates(80)
                        .build());

        store.ensureCollectionExists();
        store.add(
                "documentdb-overview",
                Embedding.from(documentVector),
                TextSegment.from(
                        "DocumentDB provides a MongoDB-compatible API.",
                        Metadata.from(Map.of("category", "database", "published", true))));

        var request = new EmbeddingSearchRequest(
                Embedding.from(queryVector),
                4,
                0.75,
                new IsEqualTo("category", "database"));
        return store.search(request).matches();
    }
}

File: examples/csharp/VectorStoreExample.cs
using DocumentDB.LangChain;
using Microsoft.Extensions.VectorData;
using MongoDB.Driver;

namespace Examples;

public static class VectorStoreExample
{
    public static async Task<IReadOnlyList<VectorSearchResult<KnowledgeRecord>>> RunAsync(
        IMongoDatabase database,
        ReadOnlyMemory<float> documentEmbedding,
        ReadOnlyMemory<float> queryEmbedding,
        CancellationToken cancellationToken = default)
    {
        var store = new DocumentDBVectorStore(database, new DocumentDBVectorStoreOptions
        {
            DiskAnnMaxDegree = 32,
            DiskAnnBuildCandidates = 50,
            DiskAnnSearchCandidates = 80,
        });
        VectorStoreCollection<string, KnowledgeRecord> collection =
            store.GetCollection<string, KnowledgeRecord>("documents");

        await collection.EnsureCollectionExistsAsync(cancellationToken);
        await collection.UpsertAsync(new KnowledgeRecord
        {
            Id = "documentdb-overview",
            Text = "DocumentDB provides a MongoDB-compatible API.",
            Category = "database",
            Embedding = documentEmbedding,
        }, cancellationToken);

        var results = new List<VectorSearchResult<KnowledgeRecord>>();
        var options = new Microsoft.Extensions.VectorData.VectorSearchOptions<KnowledgeRecord>
        {
            Filter = record => record.Category == "database",
            ScoreThreshold = 0.75,
        };
        await foreach (var result in collection.SearchAsync(queryEmbedding, 4, options, cancellationToken))
        {
            results.Add(result);
        }
        return results;
    }

    public sealed class KnowledgeRecord
    {
        [VectorStoreKey]
        public string Id { get; set; } = string.Empty;

        [VectorStoreData(StorageName = "content")]
        public string Text { get; set; } = string.Empty;

        [VectorStoreData(StorageName = "category")]
        public string Category { get; set; } = string.Empty;

        [VectorStoreVector(
            1536,
            StorageName = "embedding",
            IndexKind = IndexKind.DiskAnn,
            DistanceFunction = DistanceFunction.CosineSimilarity)]
        public ReadOnlyMemory<float> Embedding { get; set; }
    }
}